posts/inside-rust/2019-11-23-const-if-match.md

```diff
@@ -0,0 +1,156 @@
+stabilized. See [#61635] for an example of the first step and [#64028] for an
+example of the second.
+
+Personally, I've looked forward to this feature for a long time, and I can't
+wait to start playing with it. If you feel the same, I would greatly
+appreciate if you tested the limits of this feature! Try to sneak `Cell`s and
+types with `Drop` impls into places they shouldn't be allowed, blow up the
+stack with poorly implemented recursive functions (see `gcd` above), and let
+us know if something goes horribly wrong.
+
+[const-int]: https://github.com/rust-lang/rust/issues/53718
+[#61635]: https://github.com/rust-lang/rust/issues/61635
+[#64028]: https://github.com/rust-lang/rust/pull/64028
+
+## What took you so long?
+
+[Miri], which rust uses under the hood for compile-time function evaluation,
+has been capable of this for a while now. However, rust needs to statically
+guarantee certain properties about variables in a `const`, such as whether
+they allow for interior mutability or whether they have a `Drop`
+implementation that needs to be called. For example, we must reject the
+following code since it would result in a `const` being mutable at runtime!
+
+[Miri]: https://github.com/rust-lang/miri
+
+```rust
+const CELL: &std::cell::Cell<i32> = &std::cell::Cell::new(42); // Not allowed...
+
+fn main() {
+    CELL.set(0);
+    println!("{}", CELL.get()); // otherwise this could print `0`!!!
+}
+```
+
+However, it is sometimes okay for a `const` to contain a *type* that allows
+interior mutability, as long as we can prove that the actual *value* of that
+type does not. This is particularly useful for `enum`s with a "unit variant"
+(e.g., `Option::None`).
+
+```rust
+const NO_CELL: Option<&std::cell::Cell<i32>> = None; // OK
+```
+
+A more detailed (but non-normative) treatment of the rules [for `Drop`][drop]
+and [for interior mutability][interior-mut] in a const context can be found
+on the [`const-eval`] repo.
+
+It is not trivial to guarantee properties about the value of a variable when
+complex control flow such as loops and conditionals is involved. Implementing
+this feature required extending the existing dataflow framework in rust so
+that we could properly track the value of each local across the control-flow
+graph. At the moment, the analysis is very conservative, especially when values are
+moved in and out of compound data types. For example, the following will not
+compile, even when the feature gate is enabled.
+
+```rust
+const fn imprecise() -> Vec<i32> {
+    let tuple: (Vec<i32>) = (Vec::new(),);
+    tuple.0
+}
+```
+
+Even though the `Vec` created by `Vec::new` will never actually be dropped
+inside the `const fn`, we don't detect that all fields of `tuple` have been moved
+out of, and thus conservatively assume that the drop impl for `tuple` will run.
+While this particular case is trivial, there are other, more complex ones that
+would require a more expensive solution. It is an open question how precise we
```

Comment:
Maybe note that the solution would make cases not needing it also more expensive